### ISSUE TITLE
Feature ideas for graphql-ruby 2.0

### ISSUE THREAD

**Issue by theorygeek** [rmosolgo/graphql-ruby#440]
I'm not sure if there's already a backlog for upcoming work on the gem, but I thought I'd write out a couple of ideas that I've had that could improve the library, and get some feedback.

(I labeled this as 2.0 because I _think_ these are all breaking changes.)

## Passing arguments to resolver methods
Normally, field resolvers have access to three things:
- Backing-object for the type (eg, ActiveRecord model, hash, something similar)
- Arguments for the field
- Query context

But when you use the `property:` argument when declaring the field, none of these parameters are passed to the method being called:

```ruby
# some_method doesn't get the arguments or the context
field :some_field, !types.String, property: :some_method
```

But it'd be nice to be able to access those things from within the method. I can think of two ways this could work. First, they could be passed as-is, nothing fancy:
```ruby
# MethodResolve class
def call(obj, args, ctx)
  obj.public_send(@method_name, args, ctx)
end
```

Second, we could could actually pass in the arguments from the field as keyword arguments. This would feel a little more Ruby-like:
```ruby
# Example field definition:
field :some_field, !types.String, property: :some_method do
  argument :some_arg, !types.Int
  argument :some_other_arg, !types.String
end

# Example method:
def some_method(some_arg:, some_other_arg:)
end

# Or maybe the method takes context *and* arguments:
def some_method(ctx, some_arg:, some_other_arg:)
end
```

## Camelizing fields and arguments
It seems nearly ubiquitous in the GraphQL world to use camelCase for fields and arguments. But in Ruby, it's equally ubiquitous to use snake_case. I think these two things are so common, it probably warrants a feature in the gem to handle.

Perhaps it's something that would be configured:
```ruby
Schema.define do
  camelize true
end
```

If that's enabled, all fields and arguments would appear to be camelized to the outside world, but would be snake_case everywhere in application code.

It might be possible to do this right now, with the instrumentation API - I'm not sure. (If someone has solved this, it'd be awesome to see the code!) But it seems like it'd bubble into various parts of the library code (NameResolve, HashKeyResolve, for example).

Ideally, we'd be able to avoid situations like this:
```ruby
field :someField, !types.String, property: :some_field
```

## Helper methods on GraphQL types
Preface: Is this an anti-pattern?

Sometimes, the object that is behind a GraphQL type doesn't contain very much data, and what you really need is the functionality. It'd be nice if you could define helper methods on the object type itself, that you could call from your resolvers:

```ruby
SomeType = GraphQL::ObjectType.define do
  def load_related_data(obj)
    SomeLoader.for(SomeOtherType).load(obj.some_other_type_id)
  end

  field :some_field, !types.Int do
    resolve ->(obj, args, ctx) { load_related_data(obj).then { |data| data.do_something } }
  end

 field :some_other_field, !types.Int do
    resolve ->(obj, args, ctx) { load_related_data(obj).then { |data| data.do_something_else } }
  end
end
```

(Editorial note: Maybe this works already; I've always _assumed_ it doesn't.)

**Comment by rmosolgo** [comment]
Thanks for these suggestions! I love hearing where other folks' use of GraphQL is leading them.

I'm still buzzing from that [last Rich Hickey talk]() and as far as I can tell, these all fall under the category of "providing more", and thus consist of _growth_, not _breakage_!

Here are my thoughts on the features above: 


## Helper methods 

This is a freebie, it already works :D 

```ruby 
f = GraphQL::Field.define do
  def x
    :x
  end

  resolve ->(o, a, c) { x }
end
f.resolve(nil, nil, nil)
# => :x
```

Personally, I wouldn't use it because I have no idea how I'd ever test that method! 

## Better resolvers 

I've been thinking about increasing the number of ways you can register resolvers too, mostly for performance reasons. These are right up the same alley! I can imagine implementing them with new keywords, eg: 

```ruby 
# calls `post.comments(args, ctx)`
field :comments, types[CommentType], method_call: :comments 
# calls `post.comments(ctx, **args)`
field :comments, types[CommentType, keyword_call: :comments
# calls `post.comments(**args)`
field :comments, types[CommentType, splat_call: :comments
```

Maybe not those names exactly, but, get my drift?

## Camelizing 

It seems like `camelize(true)` could exist alongside the current API, right? The default is `camelize(false)` which would be the current behavior. 

I don't think this is really possible with instrumentation (field arguments are possible, but not object field names, at least not in any pretty way 😆). 

## Backlog? 

> backlog for upcoming work

Kind of the `## Todo` section in the readme, plus open issues 😬 

I'd be open to some more formal "roadmap" but I have no idea where to even start :S

**Comment by cjoudrey** [comment]
> I'd be open to some more formal "roadmap" but I have no idea where to even start :S

Would it make sense to move all the todo items in the README to their own issue that way we can provide more context than a single line for each one?

We could also tag specific issues as `2.0` or use a Milestone. I'm assuming stuff in `2.0` would mostly be breaking changes otherwise we could do them _right now_.

**Comment by rmosolgo** [comment]
That sounds good, I'll move those items when I get a chance

**Comment by theorygeek** [comment]
>I've been thinking about increasing the number of ways you can register resolvers too, mostly for performance reasons.

I'm curious, what kinds of performance gains do you think we could get?

**Comment by rmosolgo** [comment]
Here's my eye on perf gains: https://github.com/rmosolgo/graphql-ruby/issues/458

**Comment by rmosolgo** [comment]
@cjoudrey thanks for the suggestion, I've moved my TODOs to github issues!

**Comment by theorygeek** [comment]
I think all of the suggestions I raised here will be addressed in 1.8 🎉
